go/chat/crypto_doc.md

```diff
@@ -0,0 +1,205 @@
+- Encrypt the signed header with another random nonce.
+- Send the header ciphertext and the body ciphertext to the server, along with
+  both of their nonces.
+
+The contents of the header aren't secret from the server, and don't necessarily
+require encryption here. The reason for encrypting the signed header is instead
+that the signature *itself* is private. Even though the server knows who's
+talking to whom, because it's delivering all the messages, it's better that it
+can't *prove* what it knows.
+
+The purpose of the signing step is to prevent participants in the chat from
+impersonating one another. Authenticated encryption is already enough to
+prevent people outside the chat from forging messages, but because all the
+participants share the encryption key, it's not enough to distinguish one
+participant from another.
+
+Attachment encryption is done in constant size chunks, with a single short or
+empty chunk to mark the end and detect truncation. Each chunk's nonce is 16
+random bytes (shared by all chunks) plus an 8-byte sequential counter (unique
+to each chunk). This prevents nonce reuse, chunk reordering, and chunk swapping
+between messages. Encryption is layered on top of signing to hide the
+signature, and the signature includes the encryption key as associated data to
+prevent the encryption layer from being swapped out. This scheme is documented
+in detail [in the source
+code](https://github.com/keybase/client/blob/master/go/chat/signencrypt/codec.go).
+
+### Key Handling
+
+The details in this section are shared with KBFS and the rest of the Keybase
+app.
+
+All Keybase devices publish a [`crypto_box`](https://nacl.cr.yp.to/box.html)
+public key and a [`crypto_sign`](https://nacl.cr.yp.to/sign.html) public key
+when they're first provisioned. Those keys live in the user's signature chain,
+where they're connected to other other devices' keys by mutual signatures. Each
+of a user's identity proofs is also signed by one of these device keys and
+recorded in the signature chain. (PGP keys can participate in the signature
+chain too, but not in chat or KBFS.)
+
+A chat symmetric encryption key is 32 random bytes. It's shared by everyone in
+the chat, and it's the same key that they use to encrypt files in the private
+KBFS folder that those same people share. When a new device needs this shared
+key, another device that has it will encrypt it with the new device's public
+`crypto_box` key and then upload it to the Keybase server. Chat signing keys
+are the device-specific `crypto_sign` keys described above.
+
+Secret keys are stored on disk, in a file encrypted with `crypto_secretbox`,
+with a key derived from the user's Keybase password. While the user is logged
+in to Keybase, that key is stored on disk in plain text (see tradeoffs below),
+or in the system keyring on macOS. When the user logs out of Keybase, that key
+is deleted, and their secret keys are unreadable until the next login.
+
```

Comment:
Also say how locally cached chat messages are stored.  With a symmetric encryption key derived from the user's secret Curve25519 DH key.